src/python/deepgnn/graph_engine/snark/decoders.py

```diff
@@ -43,6 +43,219 @@ def decode(self, line: str) -> Iterator[Tuple[int, int, int, float, list]]:
+
+    Init Parameters Format Example
+        Same graph as the previous example, just with defaults specified.
+        ```
+        EdgeListDecoder(
+            default_node_type=1,
+            default_node_weight=.5,
+            default_node_feature_types=["int32", "float32"],
+            default_node_feature_lens=[[3],[2]],
+            default_edge_type=0,
+            default_edge_weight=.5,
+            default_edge_feature_types=["uint8"],
+                default_edge_feature_lens=[[3, 0]],
+        )
+        ```
+        graph.csv
+        ```
+        0,-1,1,1,1,1.1,1.1
+        0,1,0,4,10,1,1,1
+        1,-1,1,1,1,1.1,1.1
+        1,0,0,4,10,1,1,1
+        ```
+
+    Delimiters
+        "," is the default column delimiter, it can be overriden with the delimiter parameter.
+        "/" is the default sparse features length delimiter, it can be overriden with the
+            length_delimiter parameter.
+        "\" is the escape for the delimiter in "binary" features.
+    """
+
+    def __init__(
+        self,
+        default_node_type: Optional[int] = None,
+        default_edge_type: Optional[int] = None,
+        default_node_weight: Optional[float] = None,
+        default_edge_weight: Optional[float] = None,
+        default_node_feature_types: Optional[List[Optional[str]]] = None,
+        default_edge_feature_types: Optional[List[Optional[str]]] = None,
+        default_node_feature_lens: Optional[List[Optional[List[int]]]] = None,
+        default_edge_feature_lens: Optional[List[Optional[List[int]]]] = None,
+        delimiter=",",
+        length_delimiter="/",
+    ):
+        """Initialize the Decoder."""
+        super().__init__()
+        self.node_type = default_node_type
+        self.node_weight = default_node_weight
+        self.node_feature_types = (
+            default_node_feature_types if default_node_feature_types is not None else []
+        )
+        self.node_feature_lens = (
+            default_node_feature_lens if default_node_feature_lens is not None else []
+        )
+        self.n_node_feature = len(self.node_feature_types)
+
+        self.edge_type = default_edge_type
+        self.edge_weight = default_edge_weight
+        self.edge_feature_types = (
+            default_edge_feature_types if default_edge_feature_types is not None else []
+        )
+        self.edge_feature_lens = (
+            default_edge_feature_lens if default_edge_feature_lens is not None else []
+        )
+        self.n_edge_feature = len(self.edge_feature_types)
+
+        self.delimiter = delimiter
+        self.length_delimiter = length_delimiter
+        self.escape = r"\\"[0]
+
+    def _get_feature(
+        self, key: str, length: List[int], data: Iterator
+    ) -> Union[np.ndarray, Tuple[np.ndarray, np.ndarray], None]:
+        def _str_get(item, idx):
+            if not len(item):
+                return ""
+            return item[idx]
+
+        try:
```

Comment:
I still don't understand what this code does:
1. Why do you want to convert to int and immediately throw an exception? and later use key as a `dtype` to create arrays. Both strings `binary` and `some other random string` will raise `ValueError` if you'll try to convert it to int.
2. Where do you check `Expected feature vector key is str.` message in tests? If the error is silently swallowed inside the code, why do we need it?